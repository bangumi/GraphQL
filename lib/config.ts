import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import url from 'node:url';

import type { RedisOptions } from 'ioredis';

export const production = process.env.NODE_ENV === 'production';
export const testing = process.env.NODE_ENV === 'test';

export const projectRoot = url.fileURLToPath(new URL('..', import.meta.url));
export const pkg = JSON.parse(
  fs.readFileSync(path.resolve(projectRoot, 'package.json'), 'utf8'),
) as { version: string };

export const redisPrefix = `graphql-${pkg.version}`;

const u = url.parse(process.env.REDIS_URI ?? 'redis://127.0.0.1:3306/0');

const [username, password] = (u.auth ?? '').split(':', 2);

export const redisOption = {
  host: u.hostname ?? '127.0.0.1',
  port: u.port ? Number.parseInt(u.port) : 3306,
  db: u.pathname ? Number.parseInt(u.pathname.slice(1)) : 0,
  username: username,
  password: password,
  lazyConnect: true,
} satisfies RedisOptions;
