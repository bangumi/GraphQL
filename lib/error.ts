import type { FastifyError } from '@fastify/error';
import { createError } from '@fastify/error';
import httpCodes, { StatusCodes } from 'http-status-codes';

/** Type helper to make message argument as required */
export interface SingleMessageErrorConstructor {
  new (msg: string): FastifyError & { statusCode: number };
  (msg: string): FastifyError & { statusCode: number };
}

export const BadRequestError: SingleMessageErrorConstructor = createError(
  'BAD_REQUEST',
  '%s',
  StatusCodes.BAD_REQUEST,
);
export const NotFoundError: SingleMessageErrorConstructor = createError(
  'NOT_FOUND',
  '%s not found',
  httpCodes.NOT_FOUND,
);
export const UnexpectedNotFoundError: SingleMessageErrorConstructor = createError(
  'UNEXPECTED_NOT_FOUND',
  '%s not found',
  500,
);

export class UnimplementedError extends Error {
  constructor(msg: string) {
    super('TODO: ' + msg);
  }
}

export class UnreachableError extends Error {}
