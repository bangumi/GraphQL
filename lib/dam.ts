/** 敏感词过滤 */
import RE2 from 're2';

import { fileConfig } from './config';

const controlPattern = /[\u200B-\u200F\u202A-\u202E\uFEFF]/;

interface Option {
  nsfw_word?: string;
  disabled_words?: string;
  banned_domain?: string;
}

export class Dam {
  private readonly nsfwWord: RE2 | undefined;
  private readonly disableWord: RE2 | undefined;
  private readonly bannedDomain: RE2 | undefined;

  constructor({ banned_domain, disabled_words, nsfw_word }: Option) {
    if (nsfw_word) {
      this.nsfwWord = new RE2(nsfw_word);
    }

    if (disabled_words) {
      this.disableWord = new RE2(disabled_words);
    }

    if (banned_domain) {
      this.bannedDomain = new RE2(banned_domain);
    }
  }

  needReview(text: string): boolean {
    if (text === '') {
      return false;
    }

    if (!this.disableWord) {
      return false;
    }

    return this.disableWord.test(text.toLowerCase());
  }

  censoredWords(text: string): boolean {
    if (this.disableWord?.test(text)) {
      return true;
    }

    if (this.bannedDomain != undefined) {
      return this.bannedDomain.test(text);
    }

    return false;
  }

  static allCharacterPrintable(s: string): boolean {
    return !controlPattern.test(s);
  }
}

export const dam = new Dam(fileConfig);
