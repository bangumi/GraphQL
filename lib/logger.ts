import process from 'node:process';

import { pino } from 'pino';

import { production, VERSION } from './config';

export const logger = pino({
  level: 'info',
  base: production ? { pid: process.pid, version: VERSION } : undefined,
  timestamp() {
    return `,"time":"${new Date().toISOString()}"`;
  },
  /** 使用 config.ts 的 production 变量会导致循环 import，所以直接从环境变量中读取 */
  transport: production ? undefined : { target: 'pino-pretty', options: { colorize: true } },
  formatters: {
    level(level) {
      return { level };
    },
  },
});
