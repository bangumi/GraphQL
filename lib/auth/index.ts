import { createError } from '@fastify/error';

import prisma from '../prisma';
import type { Permission } from './permission';
import { getPermission } from './permission';

export interface IUser {
  ID: number;
  username: string;
  nickname: string;
  img: string;
}

export interface IAuth {
  login: boolean;
  allowNsfw: boolean;
  permission: Readonly<Permission>;
  user: null | IUser;
}

export const TokenNotValidError = createError(
  'TOKEN_INVALID',
  "can't find user by access token",
  401,
);

export async function byHeader(key: string | string[] | undefined): Promise<IAuth> {
  if (Array.isArray(key)) {
    throw new HeaderInvalid("can't providing multiple access token");
  }
  if (!key) {
    return { login: false, permission: {}, allowNsfw: false, user: null };
  }
  if (!key.startsWith(tokenPrefix)) {
    throw new HeaderInvalid('authorization header should have "Bearer ${TOKEN}" format');
  }
  return await byToken(key.slice(tokenPrefix.length));
}

export async function byToken(access_token: string | undefined): Promise<IAuth> {
  if (!access_token) {
    return {
      user: null,
      login: false,
      permission: {},
      allowNsfw: false,
    };
  }
  const token = await prisma.chii_oauth_access_tokens.findFirst({
    where: { access_token, expires: { gte: new Date() } },
  });

  if (!token) {
    throw new TokenNotValidError();
  }

  const user = await prisma.chii_members.findFirst({
    where: { uid: Number.parseInt(token.user_id!) },
  });

  if (!user) {
    throw new Error(
      'missing user, please report a issue at https://github.com/bangumi/GraphQL/issues',
    );
  }

  return {
    user: {
      ID: user.uid,
      nickname: user.nickname,
      username: user.username,
      img: user.avatar,
    },
    login: true,
    permission: await getPermission(user.groupid),
    allowNsfw: user.regdate - Date.now() / 1000 <= 60 * 60 * 24 * 90,
  };
}

export const HeaderInvalid = createError('AUTHORIZATION_INVALID', '%s', 401);
export const tokenPrefix = 'Bearer ';
