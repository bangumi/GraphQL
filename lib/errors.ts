import { createError } from '@fastify/error';
import httpCodes from 'http-status-codes';

export const NotFoundError = createError('NOT_FOUND', '%s not found', httpCodes.NOT_FOUND);

export const UnexpectedNotFoundError = createError('UNEXPECTED_NOT_FOUND', '%s not found', 500);
