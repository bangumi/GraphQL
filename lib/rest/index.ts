import * as url from 'node:url';
import * as path from 'node:path';

import type { FastifyInstance } from 'fastify';
import type { TypeBoxTypeProvider } from '@fastify/type-provider-typebox';
import Cookie from '@fastify/cookie';

import { projectRoot } from '../config';
import type { IAuth } from '../auth';
import * as auth from '../auth';
import { ErrorRes, User } from '../types/user';
import { walk } from '../utils';
import type { App } from './type';
import prisma from '../prisma';
import type { IUser } from '../orm';

const setups: ((app: App) => void)[] = [];

for await (const file of walk(path.resolve(projectRoot, 'lib/rest/api'))) {
  if (!file.endsWith('.ts')) {
    continue;
  }

  if (file.endsWith('.test.ts')) {
    continue;
  }

  const api = (await import(url.pathToFileURL(file).toString())) as {
    setup?: (app: FastifyInstance) => void;
  };
  const setup = api.setup;
  if (setup) {
    setups.push(setup);
  }
}

export function setup(app: FastifyInstance) {
  app.addSchema(User);
  app.addSchema(ErrorRes);

  void app.register(Cookie, {
    hook: 'preHandler', // set to false to disable cookie autoparsing or set autoparsing on any of the following hooks: 'onRequest', 'preParsing', 'preHandler', 'preValidation'. default: 'onRequest'
    parseOptions: {}, // options for parsing cookies
  });

  app.addHook('preHandler', async (req) => {
    if (req.headers.authorization) {
      const a = await auth.byHeader(req.headers.authorization);
      req.user = a.user;
      req.auth = a;

      return;
    }

    if (req.cookies.sessionID) {
      const session = await prisma.chii_os_web_sessions.findFirst({
        where: { key: req.cookies.sessionID },
      });
      if (!session) {
        return;
      }
      const a = await auth.byUserID(session.user_id);

      req.user = a.user;
      req.auth = a;
    }
  });

  const server = app.withTypeProvider<TypeBoxTypeProvider>();
  for (const setup of setups) {
    setup(server);
  }

  return app;
}

declare module 'fastify' {
  interface FastifyRequest {
    auth: IAuth;
    user: IUser | null;
  }
}
