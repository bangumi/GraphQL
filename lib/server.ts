import * as fs from 'node:fs';
import * as path from 'node:path';

import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import mercurius from 'mercurius';
import AltairFastify from 'altair-fastify-plugin';
import swagger from '@fastify/swagger';
import type { OpenAPIV3 } from 'openapi-types';

import { schema } from './graphql/schema';
import type { Context } from './graphql/context';
import prisma from './prisma';
import * as auth from './auth';
import * as rest from './rest';
import { pkg, projectRoot } from './config';
import { Security } from './openapi';

export async function createServer(opts: FastifyServerOptions = {}): Promise<FastifyInstance> {
  const server = fastify(opts);

  await server.register(mercurius, {
    schema,
    path: '/v0/graphql',
    graphiql: false,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      return {
        prisma,
        auth: await auth.byHeader(request.headers.authorization),
      };
    },
  });

  await server.register(AltairFastify, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  const swaggerUI = fs.readFileSync(path.join(projectRoot, './lib/swagger.html'));

  server.get('/v0.5/', (_, res) => {
    void res.type('text/html').send(swaggerUI);
  });

  server.get('/v0.5/openapi.json', () => {
    return server.swagger();
  });

  const openapi: Partial<OpenAPIV3.Document> = {
    info: {
      version: pkg.version,
      title: 'hello',
    },
    components: {
      securitySchemes: {
        [Security.OptionalHTTPBearer]: {
          type: 'http',
          description:
            '不强制要求用户认证，但是可能看不到某些敏感内容内容（如 NSFW 或者仅用户自己可见的收藏）',
          scheme: 'Bearer',
        },
        [Security.HTTPBearer]: {
          type: 'http',
          description: '需要使用 access token 进行认证',
          scheme: 'Bearer',
        },
      },
    },
  };

  await server.register(swagger, {
    openapi,
  });

  await server.register(rest.setup, { prefix: '/v0.5' });

  return server;
}
