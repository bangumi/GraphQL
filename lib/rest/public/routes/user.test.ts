import { expect, test } from 'vitest';

import { createServer } from 'app/lib/server';

const app = await createServer();

test('should get user', async () => {
  const res = await app.inject({ url: '/v0.5/users/382951' });
  expect(res.json()).toMatchSnapshot();
  expect(res.statusCode).toBe(200);
});

test('should get user', async () => {
  const res = await app.inject({ url: '/v0.5/users/non-existing-user' });
  expect(res.json()).toMatchSnapshot();
  expect(res.statusCode).toBe(404);
});
