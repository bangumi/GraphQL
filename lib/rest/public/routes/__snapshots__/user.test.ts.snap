// Vitest Snapshot v1

exports[`should get user 1`] = `
{
  "id": 382951,
  "nickname": "树洞酱",
  "username": "382951",
}
`;

exports[`should get user 2`] = `
{
  "code": "NOT_FOUND",
  "error": "Not Found",
  "message": "user not found",
  "statusCode": 404,
}
`;
